### DISCUSSION TITLE
bulk edit of due *times* (not dates)

### DISCUSSION THREAD

**Discussion by ebonakdarian** [ucfopen/canvasapi#542] [Q&A]
Is there a way to set default due *times* for a class with this API? The Canvas interface only let's you set on the hour, I need on the 30 minute mark.

I looked through the API docs, but didn't see a way to set the due time for each assignment per class, perhaps I missed it?

I.e., I don’t see a bulk way to edit all my **9:30 AM** due times form a previous semester to say **12:30 PM** for this semester once they have been imported (dates are adjusted ok for the most part). Each assignment has two due dates/time (regular and late), say 20 assignments per semester, 40 new dates/time to be set, 3 courses, that’s a 120 time this has to be done (and without errors). Tedious and waste of time, so if I could write a short Python script to make this change it would be great.

**Comment by Thetwam** [comment]
> Is there a way to set default due times for a class with this API?

For default due times, there's a parameter `default_due_time` that you can send as part of updating course settings.

[Canvas docs for "Update course settings" endpoint](https://canvas.instructure.com/doc/api/courses.html#method.courses.update_settings)

[CanvasAPI docs for `Course.update_settings`](https://canvasapi.readthedocs.io/en/stable/course-ref.html#canvasapi.course.Course.update_settings)

Example:

```python
course.update_settings(default_due_time="23:59:59")
```

---

If you're looking to modify assignments directly, [`Assignment.edit`](https://canvasapi.readthedocs.io/en/stable/assignment-ref.html#canvasapi.assignment.Assignment.edit) allows you to change the due date, which includes the time. I don't see a direct bulk option, but you can loop over all the assignments in a course:

```python
assignments = course.get_assignments()
for assignment in assignments:
    assignment.edit(assignment={'due_at': '2022-08-27T23:59:59Z'})
```

---

Hopefully that gives you some ideas, let us know how it goes!

**Reply by ebonakdarian** [reply]
Thank you for the pointers, much appreciated, but is there a way to do this without setting the dates? The dates are all correct, all I want to do is just shift the due time, e.g., 9:30 AM to 12:30 PM. Unless I access and save the existing date first, then set the new time with the date I just pulled out. Is that something that's possible?

Of course ideally I'd have this functionality:

assignment.edit(assignment={'due_at': '*23:59:59Z'})

I.e., assignment.edit(assignment={'due_at': '*12:30:00Z'})

but I don't think that's available.

**Reply by jessemcbride** [reply]
Hi @ebonakdarian. I don't think the API supports that kind of wildcard syntax, but you can still do this fairly cleanly by skipping date-related args when modifying existing `assignment.due_at_date` objects with  `datetime.replace` or `datetime.timedelta`.

```python
# Using timedelta.replace if the exact time is known
assignment.edit(
     assignment={
           "due_at": assignment.due_at_date.replace(hour=12, minute=30, second=0)
     }
)

# Using timedelta to extend or subtract by X time
from datetime import timedelta

assignment.edit(
     assignment={
           "due_at": assignment.due_at_date + timedelta(hours=3)
     }
)
```

**Reply by ebonakdarian** [reply]
Super, thank you @jessemcbride, I'll give this a try, much appreciated.

**Reply by ebonak** [reply]
> > Is there a way to set default due times for a class with this API?
> 
> For default due times, there's a parameter `default_due_time` that you can send as part of updating course settings.
> 
> [Canvas docs for "Update course settings" endpoint](https://canvas.instructure.com/doc/api/courses.html#method.courses.update_settings)
> 
> [CanvasAPI docs for `Course.update_settings`](https://canvasapi.readthedocs.io/en/stable/course-ref.html#canvasapi.course.Course.update_settings)
> 
> Example:
> 
> ```python
> course.update_settings(default_due_time="23:59:59")
> ```

Oh, I somehow missed this message, thank you, I will give this a try too.

**Reply by ebonak** [reply]
> > Is there a way to set default due times for a class with this API?
> 
> For default due times, there's a parameter `default_due_time` that you can send as part of updating course settings.
> 
> [Canvas docs for "Update course settings" endpoint](https://canvas.instructure.com/doc/api/courses.html#method.courses.update_settings)
> 
> [CanvasAPI docs for `Course.update_settings`](https://canvasapi.readthedocs.io/en/stable/course-ref.html#canvasapi.course.Course.update_settings)
> 
> Example:
> 
> ```python
> course.update_settings(default_due_time="23:59:59")
> ```

Thanks - this works great. Now the only thing that's missing is that I can't set the "Until" time. E.g,  my class is at 10:30 AM, the assignment is due on Sept 9 at 10:30 AM (so that default now works). Students have 24 hours (until next class) at 10:30 AM - this still defaults to 11:59 PM -- ideally it would be 10:30 AM as well, but I didn't see anything in the API docs that let me set that.

But this is so much better than before, so many thanks!

**Reply by ebonak** [reply]
Hi @jessemcbride ,

Here's an update on my efforts. I am mostly able to update the times for my existing courses when I import them and adjust them to the current time for this semester. There's a problem with the time once Daylight Savings Time ends, so I need to take a look at the docs, I think there was some reference to time zones.

The other thing is that it throws an exception when the DUE and UNTIL times are set to the same value (this is when I have no late submissions). Haven't figured out what to do about this yet.

While my efforts haven't been perfect, your suggestions and code snippets have been super-useful, thank you so much for the help.

```

"""BUGS/issues with code:

after Daylight Savings Time ends, times shift, so 10:30 AM => 9:30
etc, need to fix this.  Also, program throws execptions when there's
no late submission, ie DUE/UNTIL are set to the same value .. not sure
what to do with this, could just ignore this but report on it.

"""

HOUR_OFFSET = 5
DUE_HOUR =   ... # varies per course

for c in COURSES:
    course = canvas.get_course(c)
    # Access the course's name
    print(course.name)

    assigs = course.get_assignments()
    for a in assigs:
        # make available
        if a.unlock_at is not None:
            a.edit( assignment={ "unlock_at": a.unlock_at_date.replace(hour=DUE_HOUR+HOUR_OFFSET, minute=00, second=0) } )

        # due time
        if a.due_at is not None:
            a.edit( assignment={ "due_at": a.due_at_date.replace(hour=DUE_HOUR+HOUR_OFFSET, minute=30, second=0) } )

        # submit until
        if a.lock_at is not None:
            a.edit( assignment={ "lock_at": a.lock_at_date.replace(hour=DUE_HOUR+HOUR_OFFSET, minute=30, second=0) } )


```
